src/clr-angular/data/datagrid/render/main-renderer.ts

```diff
@@ -100,6 +109,15 @@ export class DatagridMainRenderer<T = any> implements AfterContentInit, AfterVie
     }
   }
 
+  private setupColumns() {
+    this.columnsService.columns = this.headers.toArray().map(() => new BehaviorSubject<DatagridColumnState>({}));
```

Comment:
I think you can directly transform `QueryList` with `map()` without turning it into array.